### PR TITLE
Prevent transient errors during MicroK8s API server restarts from causing failures

### LINKED ISSUE #25: MicroK8s provider with channel 1.19 fails to bootstrap
### Summary

Using the `microk8s` provider with `1.19` channel consistently fails while bootstrapping, after enabling storage and dns:

- https://github.com/MiaAltieri/kfserving-operator/runs/4090356818?check_suite_focus=true#step:3:123
- https://github.com/neoaggelos/registry-charm/runs/4090959667?check_suite_focus=true#step:3:123

This Launchpad bug seems relevant, https://bugs.launchpad.net/juju/+bug/1937282.

Things are better in newer microk8s versions (1.22 tested only). The issue seems to be that the apiserver takes a long time before accepting new connections after restarting, so the kubectl call immediately fails.

### Proposed fix

Retry the kubectl calls with a back-off, to account for the api server taking longer than expected to come up.

A quickfix can be seen in https://github.com/neoaggelos/actions-operator/blob/19a8cb6692d5c25f8c67eb0f8735a89ee5267baf/dist/bootstrap/index.js#L1597-L1605

which did result in the bootstrapping process succeeding:

- https://github.com/MiaAltieri/kfserving-operator/runs/4091857022?check_suite_focus=true#step:3:125

### PR DESCRIPTION
**Pull request by neoaggelos**
### Summary

Closes #25

### Changes

- Retry failed kubectl calls after microk8s restart with a back-off
- This PR contains a number of unrelated changes that occured by building the typescript before making any source changes. These are separated in the first commit only. If requested, I can remove them from the PR.

### CODE REVIEW

File: src/bootstrap/index.ts [hunk 1]
```diff
@@ -13,29 +13,40 @@ async function exec_as_microk8s(cmd, options = {}) {
...
-        }
-        // sleep 10s
-        await new Promise(resolve => setTimeout(resolve, 10000));
-    }
-    console.log("Found test SA token; removing");
+    if (! await retry_until_zero_rc("microk8s kubectl get secrets | grep -q test-sa-token-", 12, 10000)) {
+        core.setFailed("Timed out waiting for test SA token");
+        return false;
+    };
+    core.info("Found test SA token; removing");
```
Comment (neoaggelos) [inline src/bootstrap/index.ts:49]: This is the only relevant part of the PR that needs review, all changes in `dist/` are generated by `npm run build`

File: src/bootstrap/index.ts [hunk 2]
```diff
@@ -13,29 +13,40 @@ async function exec_as_microk8s(cmd, options = {}) {
...
+        core.setFailed("Timed out waiting for CoreDNS");
+        return false;
+    };
+    if (! await retry_until_zero_rc("microk8s kubectl -n kube-system rollout status deployment/hostpath-provisioner", 12, 10000)) {
+        core.setFailed("Timed out waiting for Storage");
+        return false;
+    };
     await exec_as_microk8s("microk8s kubectl create serviceaccount test-sa");
-    let rc = 0;
-    console.log("Waiting for test SA token...");
```
Comment (addyess) [inline src/bootstrap/index.ts:25]: @neoaggelos did we lose this status line?

**Review by addyess — COMMENTED** [review]
LGTM.  Only one little nitpick

File: src/bootstrap/index.ts (hunk 2 above)
Comment (johnsca) [reply src/bootstrap/index.ts:25]: We did, but there's enough other output around that area that I don't think it's worth worrying about.

**Review by johnsca — APPROVED** [review]

**Pull request merged**